### PR TITLE
feat: ReferencePolicies replaced by ReferenceGrants

### LINKED ISSUE #2775: ReferenceGrant API is supported by the controller
### Problem Statement

With the `0.5.0` release of the Gateway API, the `ReferencePolicy` has been deprecated in favor of the `ReferenceGrant` API. This was just a renaming of the API, as all the fields have been kept the same. The `ReferencePolicy` controller must be replaced by a new `ReferenceGrant` controller.

### Acceptance Criteria

- [ ] The `ReferencePolicy` controller is replaced by a new `ReferenceGrant` controller.

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:
The Gateway API conformance tests now require the `ReferenceGrant` API to be supported by the controller. This API is intended to supersede the `ReferencePolicy` API starting from the next API minor release, as it is stated in the API: https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/referencepolicy_types.go#L32-L33. This code intent is to watch the `ReferenceGrants` in the cluster and populate the kong store to allow their usage for filtering the allowed routes.

**Which issue this PR fixes**:
Fixes #2775 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/store/fake_store_test.go [hunk 1]
```diff
@@ -858,6 +858,32 @@ func TestFakeStoreReferencePolicy(t *testing.T) {
...
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "bar",
+			},
+			Spec: gatewayv1alpha2.ReferenceGrantSpec{},
+		},
+	}
+	store, err := NewFakeStore(FakeObjects{ReferenceGrants: grants})
+	require.Nil(err)
```
Comment (pmalek) [inline internal/store/fake_store_test.go:880]:
```suggestion
	require.NoError(err)
```

File: internal/store/fake_store_test.go [hunk 2]
```diff
@@ -858,6 +858,32 @@ func TestFakeStoreReferencePolicy(t *testing.T) {
...
+	require.NotNil(store)
+	routes, err := store.ListReferenceGrants()
+	assert.Nil(err)
```
Comment (pmalek) [inline internal/store/fake_store_test.go:883]:
```suggestion
	assert.NoError(err)
```

File: internal/dataplane/parser/translate_utils.go [hunk 3]
```diff
@@ -86,6 +86,7 @@ func isRefAllowedByPolicy(
 // ReferencePolicy with those Tos and the input From.
```
Comment (pmalek) [inline internal/dataplane/parser/translate_utils.go:86]: I believe this godoc has to be updated as well, right?

**Review by pmalek — COMMENTED** [review]
Some minor nits only.

File: internal/dataplane/parser/translate_utils.go (hunk 3 above)
Comment (mlavacca) [reply internal/dataplane/parser/translate_utils.go]: Yep, I updated it.

**Review by pmalek — COMMENTED** [review]
So, if I get this right, now a user can specify policies using either `ReferencePolicy` or `ReferenceGrant`. Is that expected? Do we intend to drop `ReferencePolicy` support at some point?

**Comment by mlavacca** [comment]
> So, if I get this right, now a user can specify policies using either `ReferencePolicy` or `ReferenceGrant`. Is that expected? Do we intend to drop `ReferencePolicy` support at some point?

Correct, a user can now specify policies using both APIs. Since `ReferenceGrant` is meant as a replacement for `ReferencePolicy`, We should expect to drop its support in the future.

File: internal/controllers/gateway/referencegrant_controller.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &gatewayv1alpha2.ReferenceGrant{}},
+		&handler.EnqueueRequestForObject{},
+	)
+}
```
Comment (shaneutt) [inline internal/controllers/gateway/referencegrant_controller.go:105]: nit: every other controller puts it's setup function above the methods, usually right under the reconciler struct definition. Ideally for consistency we should do that here as well.

**Review by shaneutt — CHANGES_REQUESTED** [review]
So actually there's a bit of missing context here. For `v0.5.0` of the Gateway API we at one point had fully removed `ReferencePolicy` (in favor of it's renamed version `ReferenceGrant`) but last minute before the release we replaced it because a specific implementation had customers on it.

This is an alpha API, and we're at an alpha level of maturity with this integration and it's behind a feature gate. It is OK for us to drop the `ReferencePolicy` altogether right now rather than waiting, reducing our code maintenance and scope.

I suggest we re-work this PR to _remove ReferencePolicy_. Ideally this PR would instead basically just be a renaming.

**Review by pmalek — APPROVED** [review]
LGTM

**Review by shaneutt — APPROVED** [review]

**Pull request merged**